Signal/src/ViewControllers/PhotoLibrary/PhotoLibrary.swift

```diff
@@ -0,0 +1,288 @@
+//
+//  Copyright (c) 2018 Open Whisper Systems. All rights reserved.
+//
+
+import Foundation
+import Photos
+import PromiseKit
+
+protocol PhotoLibraryDelegate: class {
+    func photoLibraryDidChange(_ photoLibrary: PhotoLibrary)
+}
+
+class PhotoMediaSize {
+    var thumbnailSize: CGSize
+
+    init() {
+        self.thumbnailSize = .zero
+    }
+
+    init(thumbnailSize: CGSize) {
+        self.thumbnailSize = thumbnailSize
+    }
+}
+
+class PhotoPickerAssetItem: PhotoGridItem {
+
+    let asset: PHAsset
+    let album: PhotoCollectionContents
+    let photoMediaSize: PhotoMediaSize
+
+    init(asset: PHAsset, album: PhotoCollectionContents, photoMediaSize: PhotoMediaSize) {
+        self.asset = asset
+        self.album = album
+        self.photoMediaSize = photoMediaSize
+    }
+
+    // MARK: PhotoGridItem
+
+    var type: PhotoGridItemType {
+        if asset.mediaType == .video {
+            return .video
+        }
+
+        // TODO show GIF badge?
+
+        return  .photo
+    }
+
+    func asyncThumbnail(completion: @escaping (UIImage?) -> Void) -> UIImage? {
+        album.requestThumbnail(for: self.asset, thumbnailSize: photoMediaSize.thumbnailSize) { image, _ in
```

Comment:
One change (as mentioned previously) is to pass through the desired thumbnail size.